### DISCUSSION TITLE
error: unknown type name 'bool'

### DISCUSSION THREAD

**Discussion by amreus** [jacob-carlborg/dstep#270] [Q&A] [answered]
Non-progammer and D newbie here. I thought I would ask here before reporting as a bug.

Using `dstep` v1.0.0 on Windows 10 64-bit, `dstep` is failing to convert any of the C  header files (*.h) that use a `bool` type giving the error:

 `error: unknown type name 'bool'`

Header files that do not use the `bool` type convert with no errors. 

I'm not sure where the problem lies since it seems unlikely that a fundamental type would be unknown to dstep. I assume it's something in my system or setup.

Any guidance is appreciated, thanks.

**Comment by jacob-carlborg** [comment, marked as answer]
Do you have an example of some code that DStep fails to convert? Also how you run the DStep command.

C doesn't have a type called `bool`. It has a type called `_Bool` and the `stdbool.h` header file provides an alias called `bool`. This type was added in C99. The `stdbool.h` file is a compiler internal header file which is bundled with DStep. It should work out of the box. The only reason I can think of is that DStep defaults to an older standard of C, were this type is not available. You can use the `-std=<standard>` flag to specify the C standard, see [1] for more information. DStep uses Clang under the hood, any flags that can be passed to Clang can be passed to DStep.

[1] https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-std

**Reply by amreus** [reply]
> Do you have an example of some code that DStep fails to convert? 

See below. I removed most of the other functions and struct members that were in the file.

> Also how you run the DStep command.

I download DStep from this repo: `dstep-1.0.0-win64.7z` which contains one file `dstep.exe`

I moved the file `dstep.exe` to a directory in my PATH.

I `cd` to the header directory and run `dstep example.h`

The output is: `example.h:9:3: error: unknown type name 'bool'`

I tried using `dstep -std=c99 example.h`, and also tried all of the available "c" options for -std=<> but the result was the same error.

When I added `#include <stdbool.h>` to the example.h file, it did work.



**Simplified example header:**
```
//  example.h
#pragma pack(push, 8)
#ifdef __cplusplus
extern "C" {
#endif

struct ExtensionLicense {
  /// Main flag indicating whether the extension is allowed to run or not.
  bool is_licensed;
};

int GetExtensionLicense(const char* extension_id,
    struct ExtensionLicense* out_license);

#ifdef __cplusplus
} // extern "C" {
#endif
#pragma pack(pop)
```

**Reply by amreus** [reply]
I was just browsing the README. The section on Windows mentions some requirements. It is not clear if the requirements are for running dstep or building?

If any of this matter:  I do not have LLVM installed. I do have VS2019 Community edition installed, but it is not on my PATH by default.   I also have gcc 9.3.0 installed and is on my PATH by default (installed using the `scoop` package manager.)

**Reply by jacob-carlborg** [reply]
> When I added #include <stdbool.h> to the example.h file, it did work.

That's actually expected, now when I think about it. The problem with C and the preprocessor is that it's designed to work on a complete translation unit. DStep on the other hand is designed to work on a single header file. What happens here is that most likely there's another header file included before `example.h` which includes `<stdbool.h>`. The declarations in `<stdbool.h>` then becomes available to `example.h` as well. In my opinion that is a broken design of the library. 

They way you can fix this is, as you mentioned, is to include `<stdbool.h>`. Or if you don't want to/can't modify the header file you can use the `-include` flag to force include a header file before the one that is currently being processed. In this case the `<stdbool.h>` is shipped with the compiler and not the SDK. In the case of DStep it's embedded inside the executable, so it's unfortunately not accessible from the file system. The simplest way is to copy the file from the DStep repository [1] and save it somewhere on your computer. There's some more information about this at the bottom on the readme (the last bullet point) [2].

[1] https://github.com/jacob-carlborg/dstep/blob/master/resources/stdbool.h
[2] https://github.com/jacob-carlborg/dstep#limitationsknown-issues

**Reply by amreus** [reply]
@jacob-carlborg Thank you for the responses.  I have managed to get things working mostly.

[This Github repo](https://github.com/amreus/sketchup-api-d-bindings) is what I am actually attempting.   Running `run-dstep.bat` should build all the .d files.

**Reply by jacob-carlborg** [reply]
I see you've manually defined the `bool` symbol. I guess that works too.

**Reply by amreus** [reply]
> I see you've manually defined the bool symbol. I guess that works too.

I could not get `-include` syntax to work, so I looked for an alternative.  Thanks again for all your help.
